testsuite/pytests/test_eprop_plasticity.py

```diff
@@ -0,0 +1,731 @@
+    nest.Connect(nrns_rec, nrns_out, params_conn_all_to_all, params_syn_out)
+    nest.Connect(nrns_out, nrns_rec, params_conn_all_to_all, params_syn_feedback)
+    nest.Connect(gen_rate_target, nrns_out, params_conn_one_to_one, params_syn_rate_target)
+
+    nest.Connect(nrns_in + nrns_rec, sr, params_conn_all_to_all, params_syn_static)
+
+    nest.Connect(mm_rec, nrns_rec_record, params_conn_all_to_all, params_syn_static)
+    nest.Connect(mm_out, nrns_out, params_conn_all_to_all, params_syn_static)
+
+    # Create input
+
+    input_spike_rate = 0.05
+    spike_probability = duration["step"] * input_spike_rate
+    dtype_in_spks = np.float32
+
+    input_spike_bools = np.random.rand(n_batch, steps["sequence"], n_in) < spike_probability
+    input_spike_bools = np.hstack(input_spike_bools.swapaxes(1, 2))
+    input_spike_bools[:, 0] = 0
+
+    sequence_starts = np.arange(0.0, duration["task"], duration["sequence"])
+    params_gen_spk_in = []
+    for input_spike_bool in input_spike_bools:
+        input_spike_times = np.arange(duration["sequence"] * n_batch)[input_spike_bool]
+        input_spike_times_all = [input_spike_times + start for start in sequence_starts]
+        params_gen_spk_in.append({"spike_times": np.hstack(input_spike_times_all).astype(dtype_in_spks).tolist()})
+
+    nest.SetStatus(gen_spk_in, params_gen_spk_in)
+
+    # Create output
+
+    def generate_superimposed_sines(steps_sequence, periods):
+        n_sines = len(periods)
+
+        amplitudes = np.random.uniform(low=0.5, high=2.0, size=n_sines)
+        phases = np.random.uniform(low=0.0, high=2.0 * np.pi, size=n_sines)
+
+        sines = [
+            A * np.sin(np.linspace(phi, phi + 2.0 * np.pi * (steps_sequence // T), steps_sequence))
+            for A, phi, T in zip(amplitudes, phases, periods)
+        ]
+
+        superposition = sum(sines)
+        superposition -= superposition[0]
+        superposition /= max(np.abs(superposition).max(), 1e-6)
+        return superposition
+
+    target_signal = generate_superimposed_sines(steps["sequence"], [1000.0, 500.0, 333.0, 200.0])
+
+    params_gen_rate_target = {
+        "amplitude_times": (np.arange(0, duration["task"], duration["step"]) + duration["step"]).tolist(),
+        "amplitude_values": np.tile(target_signal, n_iter * n_batch).tolist(),
+    }
+
+    nest.SetStatus(gen_rate_target, params_gen_rate_target)
+
+    # Simulate
+
+    nest.Simulate(duration["sim"])
+
+    # Read out recorders
+
+    events_mm_out = mm_out.get("events")
+
+    # Evaluate training error
+
+    readout_signal = events_mm_out["readout_signal"]
+    target_signal = events_mm_out["target_signal"]
+
+    readout_signal = readout_signal[steps["shift"] - 1 :]
+    target_signal = target_signal[steps["shift"] - 2 : -1]
+
+    error = (readout_signal - target_signal) ** 2
+    loss = 0.5 * np.add.reduceat(error, np.arange(0, steps["task"], steps["sequence"]))
+
+    # Verify results
+
+    loss_NEST_verification = np.array(
```

Comment:
Where does this data come from? Are you testing this NEST vs some other NEST?

Reply:
We run it once with NEST and take the obtained results as the verification. Our reasoning was that the purpose of these tests is to catch any future changes of the NEST code base that break the eprop code, or if the installation of a user that installs NEST from source is  faulty and this would effect eprop.

Reply:
That is a good idea. Could you explain this and ideally add the Git hash of the NEST code with which you created the reference data, plus information on OS, Python and NumPy versions used to produce it?

Reply:
Fixed in https://github.com/nest/nest-simulator/pull/2867/commits/2432a91c3d5c07ab51f21674b8bc73df4bbdd5ac.